### DISCUSSION TITLE
Reading a value from a function for a Cypress test run

### DISCUSSION THREAD

**Discussion by thepixelhead** [cypress-io/cypress#28182] [Questions and Help] [answered]
In a cypress test run I want to get a value in my fileWithTest.js from a function from the fileWithObjects.js.

Besides having had different error messages before I made it that the test runs completely without error messages ... but with this remark: [object Undefined]

The test part of this example code is: 
a login, 
navigation to a page, 
reading a value from a specific table cell, 
(with a funtion that is in a different file) changing page, 
type the before read value into an input field

This is the part with the "function":

fileWithObjects.js

`functionForSpecificTableCell_Month2() {
        var specificValueFromTableCell_Month2 = ''
        cy.get('.table1 > tbody > tr.tabele_row3 > td:nth-child(1)').each(($e1, index, $list) => {
            var titletext=$e1.text()
            
            if(titletext.includes('A Specific Headline')) {
                cy.get('.table1 > tbody > tr.table_row3 > td:nth-child(3)').eq(index).then(function(amount) {
                    var specificValueFromTableCell_Month2 = amount.text()
                    return specificValueFromTableCell_Month2.trim()
                })
            } 
            
        })
        return specificValueFromTableCell_Month2
}`

This is the test file that uses the "outside function":

fileWithTest.js

`describe('Test Read From Outside Function', ()=> {

    it.only('Read a value from a function that has got to be included', () => {

        cy.wait(500)
        cy.openHomePage()

        // Login
        cy.fixture('role').then(roledata => {
            const username = roledata[1].name
            const pw = roledata[1].password

            onLoginPage.userInput(username)
            onLoginPage.pwInput(pw)
        })
    
        onLoginPage.buttonEnter()
        cy.wait(500)
        cy.wait(500)

    // Navigate to page
        navigateGoTo.openTopNavi()
        navigateGoTo.visitWantedPage()

    // Get the value from a function of a different file
        let value = onPageWithTableCellObject.functionForSpecificTableCell_Month2()

        navigateGoTo.openTopNavi()
        navigateGoTo.wantedPage()

        // Insert the value from the function of the other file into an input field
        onWantedPage.Inputfiled2(toString(value))
    
    })
})`

Result:

The test runs without any error messages (all green) but this remark and value shows up in the input field:

value in input field: **[object Undefined]**

**I can not return a value from the amount function.**

With my research I did not get any results fitting to my example. Maybe my apporach is completely wrong?

I expected something like this: https://www.geeksforgeeks.org/javascript-return-statement/ Does not work for me. 

Any idea? Thank you!

**Comment by Akshat111111** [comment, marked as answer]
It appears that you are having trouble retrieving the result from the functionForSpecificTableCell_Month2 in your fileWithObjects.js. The issue arises from the asynchronous nature of Cypress instructions and the promise handling.

You are returning a value from within an asynchronous.then() block in your fileWithObjects.js. When you return a value from within this block, the value from the functionForSpecificTableCell_Month2 function is not returned. Instead, it returns from the inner function supplied to.then() without passing the value to the outer function.

To manage the asynchronous behavior, you may utilize Cypress promises appropriately. You may return a Cypress promise from your function and then handle the received value with.then() in your test file.

Here's how to make changes to your fileWithObjects.js:
`function functionForSpecificTableCell_Month2() {
    return cy.get('.table1 > tbody > tr.tabele_row3 > td:nth-child(1)').each(($e1, index, $list) => {
        var titletext = $e1.text();
        if (titletext.includes('A Specific Headline')) {
            return cy.get('.table1 > tbody > tr.table_row3 > td:nth-child(3)').eq(index).then(function(amount) {
                var specificValueFromTableCell_Month2 = amount.text().trim();
                return specificValueFromTableCell_Month2;
            });
        }
    });
}  module.exports = {
    functionForSpecificTableCell_Month2: functionForSpecificTableCell_Month2
};`  
In your test file, you can use .then() to handle the returned promise.
